### PR TITLE
Implement keyword search for projects

### LINKED ISSUE #5: Implement keyword search for projects
- [ ] Implement ability to use keywords to search for a project

### PR DESCRIPTION
**Pull request by allison-truhlar**
Adds a keyword search component to the project and ecosystem landing pages. The search functionality is implemented using fuse.js and JSON endpoints (`/projects.json` and `/ecosystems.json`) generated using Astro utilities for content collections. Merging closes #5.

One idea for an additional feature to improve the usefulness of the search - adding a "keywords" field to the project/ecosystem Markdown files where contributors can add whatever keywords they want that will be included in the search index, but will not be displayed as tags.

### CODE REVIEW

File: src/pages/projects.json.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    "use case": project.data["use case"],
+    "usage environment": project.data["usage environment"],
+    "software ecosystem": project.data["software ecosystem"],
+    "supported file types": project.data["supported file types"],
+    body: project.body,
+  }));
+}
+
+// Note - "GET" must be all uppercase. Lowercase threw an error.
+export async function GET({}) {
```
Comment (krokicki) [inline src/pages/projects.json.js:37]: Not being familiar with Astro, I'm trying to understand how this works. It looks like this file implements a backend handler which is called when the user requests /project.json, correct? Then it gets the project data from Astro and indexes it using Fuse, and returns the index?

**Comment by krokicki** [comment]
This looks great, @allison-truhlar! I checked out the branch and tried it out, and it behaves very intuitively.

File: src/pages/projects.json.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      "software type",
+      "use case",
+      "usage environment",
+      "software ecosystem",
+      "supported file types",
+      "body",
+    ],
+    projects
+  );
+  return new Response(
```
[the remark this answers is not in the document]
Comment (allison-truhlar) [reply src/pages/projects.json.js:57]: The Fuse index is calculated at build time (see my response to the comment above), and is not recalculated on each request. However, in `projectSearchResultsStore.js`, where the logic for the `projectsFuse` and `projectData` context is kept, the pre-calculated index does [need to be parsed and a new Fuse instance needs to be created](https://github.com/JaneliaSciComp/ossi-website/blob/5ac0230516707526054942e34767112649789f51/src/components/projects/stores/projectSearchResultsStore.js#L37C1-L38C78) when the projects cards on the `/projects` page or `/ecosystems` page are first mounted. This Fuse instance is then used as long as the client stays on either the `/projects` or `/ecosystems` page.

[the remark this answers is not in the document]
Comment (krokicki) [reply src/pages/projects.json.js:57]: Makes sense, thanks!

**Comment by neomorphic** [comment]
I am finding the text in the keyword search input really hard to read. Is this just on my browser? Screenshot 2024-05-07 at 9 19 20 AM

**Comment by allison-truhlar** [comment]
@neomorphic Hm interesting - that's the color of the placeholder text on Firefox. Here's what the text looks like once I start typing. I just pushed an additional commit to the branch to explicitly set the input text color as different from the placeholder text - can you test this out?
Screenshot 2024-05-07 at 9 23 50 AM

**Comment by neomorphic** [comment]
That looks much better on my screen. Must be a chrome thing

> @neomorphic Hm interesting - that's the color of the placeholder text on Firefox. Here's what the text looks like once I start typing. I just pushed an additional commit to the branch to explicitly set the input text color as different from the placeholder text - can you test this out? Screenshot 2024-05-07 at 9 23 50 AM

**Pull request merged**